internal/common/function/function.go

```diff
@@ -280,6 +281,29 @@ func CheckVeleroBackupLabels(backup *velerov1api.Backup) bool {
 	return exists && value == constant.ManagedByLabelValue
 }
 
+// DeleteVeleroBackup deletes Velero Backup object based on the NonAdminBackup object name and namespace
+func DeleteVeleroBackup(ctx context.Context, r client.Client, logger logr.Logger, nab *nacv1alpha1.NonAdminBackup) (bool, error) {
```

Comment:
Isn't the Deletion logic flawed here ? Why are we using `GenerateVeleroBackupName` function to fetch the backup to be deleted ? Isnt there a high possibility that the Generated Name won't match the actual VeleroBackup Name that we want to delete ? Shouldn't we List the Velero backups and then check the annotations of these Velero backup objects and match the nab name with the annotation value of the key `"openshift.io/oadp-nab-origin-name"` to get the VeleroBackup to be deleted ? Let me know if I am missing something ?

Reply:
Our velero backup name is *always* generated based on https://github.com/migtools/oadp-non-admin/blob/master/internal/common/function/function.go#L115-L145

```
fmt.Sprintf("%s-%s-%s", constant.VeleroBackupNamePrefix, namespace, nameHash)
```

So we have always `nab-NAMESPACE-HASH` where the HASH is sha256 from NonAdminBackup name.

VeleroBackup could get deleted only if one matches that entire 3 parts.

If we want to add extra protection we could, but then I would argue the UUID of the NonAdminBackup needs to be used from the VeleroBackup to match the UUID of the requestor.

It is currently possible, because we store this annotation within VeleroBackup:

```
openshift.io/oadp-nab-origin-uuid=<UUID_OF_NAB>
```

Let me know if I should add that protection on top of current name match.

Reply:
why not to just use 
https://github.com/migtools/oadp-non-admin/blob/43812bec2fbf7f790b9afa101c9bf395541571ff/api/v1alpha1/nonadminbackup_types.go#L52
?

Reply:
+1

Reply:
I thought about that, but with the current implementation I can:

 1. Create NAB
 2. VeleroBackup will get created
 3. Remove NAB
 
At this stage I have still can remove VeleroBackup from 2 if I create new NAB with the same name as the previous one with flag `deleteVeleroBackup: true`.

Using status will disallow me to do so, because 2 will never get created as this object will be there *forever* until sysadmin removes it. It's because Status can only be updated by controller when 2 is created.

Reply:
We need a status condition or clear message about this backup when it gets re-attached like this: it did not run a new backup. 

I think this is the most correct behavior, but it could be surprising for an end user, and we probably need to spend time on how to make this clear to the user.

Reply:
I checked `GenerateVeleroBackupName` function, and I think my fear and Shubham's wont happen: that function is not random, right? it always create same name given same input

But still, we already have that info in NAB, why not get info there instead of generating the info again?

@mpryc the problem you pointed only happens if:
- NAB is deleted manually (without setting `delete: true` which will result in Velero Backup being deleted)

(which we will say in docs that will not work)

and user will only get problem creating NAB with same name as previous if user is quicker then NAC sync controller, right?

I am ok with current implementation, just pointing out my view :smiley:

Reply:
I don't want to have corner case where for some reason NAB did not got proper update of the Status.

To be honest we may want also add some safety feature to only delete VeleroBackup if it has the UID or other linkage to VeleroBackup from NAB object, this imo is separate design ?

There needs to be one source of truth here and I've chosen our function, where it could also be Status, so either way. Function is less check if status exists as we have name and only check if VeleroBackup exists.

Reply:
The NAB will not have a status only when the NAB controller is not running right ? or if the NAB controller is running and the status update call fails ? I think this makes it even more prudent that we check for the status of NAB first:
- check if NAB has status
- lookup of VeleroBackup object from NAB.status.VeleroBackupName
- VeleroBackup is in terminal state
- then delete the VeleroBackup object and then subsequently NAB
Picking up the VeleroBackupName from NAB status at the very least makes sure that VeleroBackup exists for that NAB object and Delete Call makes sense from user perspective.

Reply:
There is a time period when the NAB was created, but Backup was not yet created. If we are in that time window there is a problem with getting proper velero backup name from status. If that happens we won't reconcile again and that object will be out of sync. That is the only additional use case which I can think of when it comes to inconsistency.

Reply:
See thats the thing, we are not allowing Delete for such cases, grabbing the name from status would not allow a delete for such cases(because there is no VeleroBackup object in terminal phase)

Reply:
Ok, more then half of myself tend to agree with you now :-) Thanks.